Add REST API docs (#3376)
scalableminds/webknossos: docs/rest_api.md

## Authentication

All API routes expect the HTTP header `X-Auth-Token` to be set to your personal webKnossos token. You can find this token in the webKnossos menu, directly below “logout”.

## Versioning

The API is subject to frequent changes. However, older versions will be supported for some time via prefixed routes.
 - To access the latest version, call the routes below with only the `/api` prefix.
 - To access a specific version, call the routes below with a prefix in the form of `/api/v1`.

New versions will be documented here, detailing the changes. Note, however, that some changes are not considered to be breaking the API and will not lead to new versions. Such changes include new optional parameters as well as new fields in the responses. The same goes for error message wording.

## Routes

### `GET /buildinfo`

#### Returns
JSON object containing information about the version of webKnossos

---
### `GET /users`

List all users for which you have read access

#### Expects
 - Optional GET parameter `isEditable=[BOOLEAN]`
   - If true: list only users for which you can change team memberships

>    - If false: list only users for which you can not change team memberships

Reviewer: cannot
```suggestion
   - If false: list only users for which you cannot change team memberships
```

   - if false: list only unfinished task annotations
 - Optional GET parameter `limit=[INT]`
   - return only the first x results (defaults to 1000)

#### Returns
 - JSON list of objects containing annotation information about the task annotations of the user, also including task and task type information

---
### `GET /users/:id/annotations`

List the explorative annotations of a uaser

#### Expects
 - In the url: `:id` id of a user
 - Optional GET parameter `isFinished=[BOOLEAN]`
   - If true: list only finished (=archived) explorative annotations
   - If false: list only unfinished (=non-archived) explorative annotations
 - Optional GET parameter `limit=[INT]`
   - return only the first x results (defaults to 1000)

#### Returns
 - JSON list of objects containing annotation information about the explorative annotations of the user

---
### `GET /teams`

List all teams that you can manage

#### Returns
 - JSON list of objects containing team information

---
### `GET /datasets`

List all datasets for which you have read access

#### Expects
 - Optional GET parameter `isEditable=[BOOLEAN]`
   - If true: list only datasets you can edit

>    - If false: list only datasets you can not edit

Reviewer: ```suggestion
   - If false: list only datasets you cannot edit
```

 - In the url: `:typ` and `:id` as described above under `GET /annotations/:typ/:id/info`
 - In the url: `:mergedTyp` and `:mergedId` as described above under `GET /annotations/:typ/:id/info`

---
### `POST /tasks`

Create tasks without attached NML files

#### Expects
 - JSON list of objects. Each has these fields:
   - `"taskTypeId"` `[STRING]`
   - `"neededExperience"` `[JSON OBJECT]` with these fields:
     - `"domain"` `[STRING]`
     - `"value"` `[INT]`
   - `"openInstances"` `[INT]`
   - `"projectName"` `[STRING]`
   - `"scriptId"` (optional) `[STRING]`
   - `"boundingBox"` (optional) `[JSON OBJECT]` with these fields:
     - `"topLeft"` `[JSON LIST OF THREE INTS]`
     - `"width"` `[INT]`
     - `"height"` `[INT]`
     - `"depth"` `[INT]`
   - `"dataSet"` `[STRING]` name of the dataset
   - `"editPosition"` `[JSON LIST OF THREE INTS]` starting coordinates
   - `"editRotation"` `[JSON LIST OF THREE INTS]` starting rotation
   - `"creationInfo"` (optional) `[STRING]` identifier to match with task creation result report
   - `"description"` (optional) `[STRING]`

#### Returns

 - JSON list with status reports about the created tasks, including their ids

#### Note

> For each attached NML file, one task is created. This request will also return status code 200 if some tasks could not be created. Failed tasks are instead encoded in the JSON response

Reviewer: I don't get this. Above it says "Create tasks without attached NML files" but here we talk about NMLs?

Create tasks with attached NML files

#### Expects
 - Form data field `"formJSON"`
   - JSON object with these fields:
     - `""`
     - `"taskTypeId"` `[STRING]`
     - `"neededExperience"` `[JSON OBJECT]` with these fields:
       - `"domain"` `[STRING]`
       - `"value"` `[INT]`
     - `"openInstances"` `[INT]`
     - `"projectName"` `[STRING]`
     - `"scriptId"` (optional) `[STRING]`
     - `"boundingBox"` (optional) `[JSON OBJECT]` with these fields:
       - `"topLeft"` `[JSON LIST OF THREE INTS]`
       - `"width"` `[INT]`
       - `"height"` `[INT]`
       - `"depth"` `[INT]`
  - Attached NML files

#### Returns

 - JSON list with status reports about the created tasks, including their ids

#### Note

For each attached NML file, one task is created. This request will also return status code 200 if some tasks could not be created. Failed tasks are instead encoded in the JSON response

---
### `POST /tasks/list`

List tasks matching search criteria

#### Expects
 - JSON object with four optional fields:
   - `"user"` `[STRING]` show only tasks on which the user with this id has worked
   - `"project"` `[STRING]` show only tasks of the project with this name

>    - `"ids"` `[JSON LIST OF STRINGS]` show only tasks with these ids

Reviewer: Inconsistent with `LIST[STRING]` used above

Author: (changed the one above)
